sound/soc/sof/intel/hda-dsp.c

```diff
@@ -412,6 +412,34 @@ static int hda_dsp_update_d0i3c_register(struct snd_sof_dev *sdev, u8 value)
 	return 0;
 }
 
+/*
+ * d0i3 streaming is enabled if all the active streams can
+ * work in d0i3 state and playback is enabled
+ */
+static bool hda_dsp_d0i3_streaming_applicable(struct snd_sof_dev *sdev)
+{
+	struct snd_pcm_substream *substream;
+	struct snd_sof_pcm *spcm;
+	bool playback_active = false;
+	int dir;
+
+	list_for_each_entry(spcm, &sdev->pcm_list, list) {
+		for_each_pcm_streams(dir) {
+			substream = spcm->stream[dir].substream;
+			if (!substream || !substream->runtime)
+				continue;
+
+			if (!spcm->stream[dir].d0i3_compatible)
+				return false;
+
+			if (dir == SNDRV_PCM_STREAM_PLAYBACK)
+				playback_active = true;
```

Comment:
I still don't quite follow the logic.

Do you want to enable D0i3 if all streams are D0i3 compatible and at least one of these is for playback?

Why this odd logic?
did you mean to check that all of them are D0i3 compatible and all of them are playback?
Can we really have a mix of D0i3 compatible streams in different directions?

Reply:
thanks @ranj063 . very glad to use your answer.  In a case that we enjoy music with deep buffer enabled and  wake up laptop with hello. It is common on windows that WOV and deep buffer are enabled.

Reply:
snd_sof_dsp_only_d0i3_compatible_stream_active was first designed for D0i3-hotwording. D0i3-streaming is for playback with deep buffer.
Pierre suggested that Conflating D0i3-hotwording and D0i3-streaming is not correct, so we introduce hda_dsp_d0i3_streaming_applicable which will check all the streams can enter d0i3 and need to enable streaming.